kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/PaginatedScanSuite.scala

```diff
@@ -0,0 +1,353 @@
+
+    assert(fileCounts.length == expectedBatchCount)
+    assert(totalFileCountsReturned == expectedFileCount)
+
+    logger.info(s"$testName: Total num batches returned in page one = ${fileCounts.length}")
+    logger.info(s"$testName: Total num Parquet Files fetched in page one = " +
+      s"$totalFileCountsReturned")
+  }
+
+  private def validatePageToken(
+      paginatedIter: PaginatedScanFilesIterator,
+      expectedLogFilePath: String,
+      expectedRowIndex: Long,
+      testName: String): Unit = {
+    val lastReadLogFilePath = PageToken.fromRow(paginatedIter.getCurrentPageToken)
+      .getLastReadLogFilePath
+    val lastReturnedRowIndex = PageToken.fromRow(paginatedIter.getCurrentPageToken)
+      .getLastReturnedRowIndex
+
+    assert(lastReadLogFilePath.endsWith(expectedLogFilePath))
+    assert(lastReturnedRowIndex == expectedRowIndex)
+
+    logger.info(s"$testName: New PageToken: lastReadLogFileName = $lastReadLogFilePath")
+    logger.info(s"$testName: New PageToken: lastReadRowIndex = $lastReturnedRowIndex")
+  }
+
+  private def runSinglePaginationTestCase(
+      testName: String,
+      tablePath: String,
+      tableVersionOpt: Optional[Long],
+      pageSize: Long,
+      expectedFileCount: Int,
+      expectedBatchCount: Int,
+      expectedLogFileName: String,
+      expectedRowIndex: Int): Unit = {
+    val paginatedScan = createPaginatedScan(
+      tablePath = tablePath,
+      tableVersionOpt = tableVersionOpt,
+      pageSize = pageSize)
+
+    val paginatedIter = paginatedScan.getScanFiles(customEngine)
+    val returnedBatchesInPage = collectPaginatedBatches(paginatedIter)
+
+    validatePageResults(returnedBatchesInPage, expectedFileCount, expectedBatchCount, testName)
+    validatePageToken(paginatedIter, expectedLogFileName, expectedRowIndex, testName)
+
+    paginatedIter.close()
+  }
+
+  private def runPaginationTestCasesForFirstPage(
+      tablePath: String,
+      tableVersionOpt: Optional[Long] = Optional.empty(),
+      testCases: Seq[SinglePageRequestTestCase]): Unit = {
+    testCases.foreach { testCase =>
+      runSinglePaginationTestCase(
+        testName = testCase.testName,
+        tablePath = tablePath,
+        tableVersionOpt = tableVersionOpt,
+        pageSize = testCase.pageSize,
+        expectedFileCount = testCase.expectedFileCount,
+        expectedBatchCount = testCase.expectedBatchCount,
+        expectedLogFileName = testCase.expectedLogFileName,
+        expectedRowIndex = testCase.expectedRowIndex)
+    }
+  }
+
+  // TODO: test call hasNext() twice
+
+  test("Pagination with single JSON file") {
+    withGoldenTable("kernel-pagination-all-jsons") { tablePath =>
+      /**
+       *  Log Segment List:
+       *  00000000000000000000.json contains 2 batches, 5 active AddFiles in total
+       *  Batch 1: 5 rows, 2 selected AddFiles
+       *  Batch 2: 3 rows, 3 selected AddFiles
+       */
+
+      val JSONFile0 = "00000000000000000000.json"
```

Comment:
nit: `JSON` to start a variable seems a little weird, since it is not `JSON_FILE_0` (which is used for static final values) -- I would suggest `jsonFile0`